### DISCUSSION TITLE
How to set cookies when using `fetchRequestHandler()`?

### DISCUSSION THREAD

**Discussion by matallui** [trpc/trpc#4792] [Q&A]
Hi!

I have a Next.js 13 (app router) which uses `fetchRequestHanlder()`.
Here's the route handler for demo purposes:

```ts
// app/api/trpc/[trpc]/route.ts
import { fetchRequestHandler } from "@trpc/server/adapters/fetch";

import { appRouter, createTRPCContext } from "@phnx/api";

import { getSession } from "./auth";

/**
 * Configure basic CORS headers
 * You should extend this to match your needs
 */
function setCorsHeaders(res: Response) {
  res.headers.set("Access-Control-Allow-Origin", "*");
  res.headers.set("Access-Control-Request-Method", "*");
  res.headers.set("Access-Control-Allow-Methods", "OPTIONS, GET, POST");
  res.headers.set("Access-Control-Allow-Headers", "*");
}

export function OPTIONS() {
  const response = new Response(null, {
    status: 204,
  });
  setCorsHeaders(response);
  return response;
}

const handler = async (req: Request) => {
  const session = await getSession();
  const response = await fetchRequestHandler({
    endpoint: "/api/trpc",
    router: appRouter,
    req,
    createContext: ({ resHeaders }) =>
      createTRPCContext({ req, auth: session, resHeaders }),
    onError({ error, path }) {
      console.error(`>>> tRPC Error on '${path}'`, error);
    },
  });

  setCorsHeaders(response);
  return response;
};

export { handler as GET, handler as POST };
```

My question is: how can I set a session cookie from the server side?

I tried using `resHeaders.set('Set-Cookie', 'session=somevalue')` as suggested in https://github.com/trpc/trpc/discussions/4226 , but that didn't seem to work. All I've tried ended up in not working altogether.

Any help would be greatly appreciated!

**Comment by martitv** [comment]
https://nextjs.org/docs/app/api-reference/functions/cookies
Could this work?

**Reply by shivangrathore** [reply]
Thank you my man, I been scrambling my brain for so long

**Comment by IDJGILL** [comment]
> Hi!
> 
> I have a Next.js 13 (app router) which uses `fetchRequestHanlder()`. Here's the route handler for demo purposes:
> 
> ```ts
> // app/api/trpc/[trpc]/route.ts
> import { fetchRequestHandler } from "@trpc/server/adapters/fetch";
> 
> import { appRouter, createTRPCContext } from "@phnx/api";
> 
> import { getSession } from "./auth";
> 
> /**
>  * Configure basic CORS headers
>  * You should extend this to match your needs
>  */
> function setCorsHeaders(res: Response) {
>   res.headers.set("Access-Control-Allow-Origin", "*");
>   res.headers.set("Access-Control-Request-Method", "*");
>   res.headers.set("Access-Control-Allow-Methods", "OPTIONS, GET, POST");
>   res.headers.set("Access-Control-Allow-Headers", "*");
> }
> 
> export function OPTIONS() {
>   const response = new Response(null, {
>     status: 204,
>   });
>   setCorsHeaders(response);
>   return response;
> }
> 
> const handler = async (req: Request) => {
>   const session = await getSession();
>   const response = await fetchRequestHandler({
>     endpoint: "/api/trpc",
>     router: appRouter,
>     req,
>     createContext: ({ resHeaders }) =>
>       createTRPCContext({ req, auth: session, resHeaders }),
>     onError({ error, path }) {
>       console.error(`>>> tRPC Error on '${path}'`, error);
>     },
>   });
> 
>   setCorsHeaders(response);
>   return response;
> };
> 
> export { handler as GET, handler as POST };
> ```
> 
> My question is: how can I set a session cookie from the server side?
> 
> I tried using `resHeaders.set('Set-Cookie', 'session=somevalue')` as suggested in #4226 , but that didn't seem to work. All I've tried ended up in not working altogether.
> 
> Any help would be greatly appreciated!

I have same question, have you found any solution yet?

**Comment by IDJGILL** [comment]
I found a fix for this. Consider these key points:

1. If you are using `unstable_httpBatchStreamLink` by any chance, remember that it does not support setting headers once the stream has begun.
2. #4226 has been answered very well and it works, but the only catch is to not use `unstable_httpBatchStreamLink` for this particular use case.
3. Make sure you pass'resHeaders''from createContext to your context, from which you can access'resHeaders' from anywhere in your procedures. Here's how I did it.

`const handler = (req: NextRequest) =>
  fetchRequestHandler({
    endpoint: "/api/trpc",
    req,
    router: appRouter,
    createContext: ({ resHeaders }) => createTRPCContext({ req, resHeaders }),
    onError:
      env.NODE_ENV === "development"
        ? ({ path, error }) => {
            console.error(
              `❌ tRPC failed on ${path ?? "<no-path>"}: ${error.message}`,
            );
          }
        : undefined,
  });

export { handler as GET, handler as POST };
`

Do let me know if this works for you. Happy codding.

**Comment by oyerindedaniel** [comment]
did you get a solution:

app\api\trpc\[trpc]\route.ts
```
const createContext = async (req: NextRequest, resHeaders: Headers) => {
  console.log('cookies---------------', cookies().get('access_refresh_token')?.value);
  console.log('header', req.headers.get('cookie'));
  console.log('header---', req.headers);
  resHeaders.set(
    'set-cookie',
    cookie.serialize('access_refresh_token', 'oyerinde-daniel', {
      httpOnly: true,
      secure: Boolean(Number(process.env.COOKIE_SECURE ?? 0)),
      sameSite: 'lax'
    })
  );
  return createTRPCContext({
    headers: req.headers,
    cookies,
    resHeaders
  });
};

const handler = (req: NextRequest) => {
  return fetchRequestHandler({
    endpoint: '/api/trpc',
    req,
    router: appRouter,
    createContext: ({ resHeaders }) => createContext(req, resHeaders),
    onError:
      env.NODE_ENV === 'development'
        ? ({ path, error }) => {
            console.error(`❌ tRPC failed on ${path ?? '<no-path>'}: ${error.message}`);
          }
        : undefined
  });
};
```

The above works, but if I try setting the header in the procedure as shown below, it does not work


```
 ctx.resHeaders.set(
        'set-cookie',
        cookie.serialize('access_refresh_token', 'oyerinde-daniel', {
          httpOnly: true,
          secure: Boolean(Number(process.env.COOKIE_SECURE ?? 0)),
          sameSite: 'lax'
        })
      );
```

**Reply by IDJGILL** [reply]
Are you using `unstable_httpBatchStreamLink` ?

**Reply by oyerindedaniel** [reply]
i am using a t3stack starter template, trying to modify it because i want to use custom auth

Do you know about t3stack if not:
I have a react.tsx, server.tsx and api\trpc[trpc]\route.ts. These are the three entry points i have (based on my understanding)

react.tsx file acts as wrapper that wraps my main layout providing trpc client features (responsible for client query). In it i have a terminating link (unstable_httpBatchStreamLink). 

react.tsx

```
'use client';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { loggerLink, unstable_httpBatchStreamLink } from '@trpc/client';
import { createTRPCReact } from '@trpc/react-query';
import { useState } from 'react';

import { type AppRouter } from '@/server/api/root';
import { getUrl, transformer } from './shared';

export const api = createTRPCReact<AppRouter>();

export function TRPCReactProvider(props: { children: React.ReactNode; cookies: string }) {
  const [queryClient] = useState(() => new QueryClient());

  const [trpcClient] = useState(() =>
    api.createClient({
      transformer,
      links: [
        loggerLink({
          enabled: (op) =>
            process.env.NODE_ENV === 'development' ||
            (op.direction === 'down' && op.result instanceof Error)
        }),
        unstable_httpBatchStreamLink({
          url: getUrl(),
          headers() {
            return {
              cookie: props.cookies,
              'x-trpc-source': 'react'
            };
          }
        })
      ]
    })
  );

  return (
    <QueryClientProvider client={queryClient}>
      <api.Provider client={trpcClient} queryClient={queryClient}>
        {props.children}
      </api.Provider>
    </QueryClientProvider>
  );
}
```
i also have server tsx (responsible for server query)
```
import 'server-only';

import { createTRPCProxyClient, loggerLink, TRPCClientError } from '@trpc/client';
import { callProcedure } from '@trpc/server';
import { observable } from '@trpc/server/observable';
import { type TRPCErrorResponse } from '@trpc/server/rpc';
import { cookies } from 'next/headers';
import { cache } from 'react';

import { appRouter, type AppRouter } from '@/server/api/root';
import { createTRPCContext } from '@/server/api/trpc';
import { transformer } from './shared';

/**
 * This wraps the `createTRPCContext` helper and provides the required context for the tRPC API when
 * handling a tRPC call from a React Server Component.
 */
const createContext = cache(() => {
  return createTRPCContext({
    headers: new Headers({
      cookie: cookies().toString(),
      'x-trpc-source': 'rsc'
    }),
    cookies
  });
});

export const api = createTRPCProxyClient<AppRouter>({
  transformer,
  links: [
    loggerLink({
      enabled: (op) =>
        process.env.NODE_ENV === 'development' ||
        (op.direction === 'down' && op.result instanceof Error)
    }),
    /**
     * Custom RSC link that lets us invoke procedures without using http requests. Since Server
     * Components always run on the server, we can just call the procedure as a function.
     */
    () =>
      ({ op }) =>
        observable((observer) => {
          createContext()
            .then((ctx) => {
              return callProcedure({
                procedures: appRouter._def.procedures,
                path: op.path,
                rawInput: op.input,
                ctx,
                type: op.type
              });
            })
            .then((data) => {
              observer.next({ result: { data } });
              observer.complete();
            })
            .catch((cause: TRPCErrorResponse) => {
              observer.error(TRPCClientError.from(cause));
            });
        })
  ]
});
```

and the [trpc]\route.ts is the one i provided in my initial message.

The issue i am having is since i am trying to setup a custom auth. I want to be able to set cookie in my auth router (procedure) but right now every attempt to set one not working. In my initial message i am passing the resHeaders from by router handler to the context, then in the procedure i am trying to set the cookie but it does not work. i also tried passing next header cookies function to the context so that i can have access to it in my procedures. It does not also work (did not expect it to work: based on this in next docs "**The cookies function allows you to read the HTTP incoming request cookies from a [Server Component](https://nextjs.org/docs/app/building-your-application/rendering/server-components) or write outgoing request cookies in a [Server Action](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) or [Route Handler](https://nextjs.org/docs/app/building-your-application/routing/route-handlers).**") 

Right now, setting the cookies only works if i try setting it outside the procedures in the route handler, before it goes into my context. "**As shown in my initial question**"

**Reply by IDJGILL** [reply]
Your code looks good to me but the only problem is in your TRPCReactProvider, It uses the new unstable_httpBatchStreamLink. It comes with amazing benefits but you can't set header.

Here's what you need to solve this issue:

```
import { httpBatchLink, loggerLink } from "@trpc/client"

export function TRPCReactProvider(props: { children: React.ReactNode; cookies: string }) {
  const [queryClient] = useState(() => new QueryClient());

  const [trpcClient] = useState(() =>
    api.createClient({
      transformer,
      links: [
        loggerLink({
          enabled: (op) =>
            process.env.NODE_ENV === 'development' ||
            (op.direction === 'down' && op.result instanceof Error)
        }),
        httpBatchLink({
          url: getUrl(),
          headers() {
            return {
              cookie: props.cookies,
              'x-trpc-source': 'react'
            };
          }
        })
      ]
    })
  );

  return (
    <QueryClientProvider client={queryClient}>
      <api.Provider client={trpcClient} queryClient={queryClient}>
        {props.children}
      </api.Provider>
    </QueryClientProvider>
  );
}
```

I just changed the unstable_httpBatchStreamLink to httpBatchLink.

To set cookies: 

```
 import { cookies } from "next/headers"

// In your procedure
cookie.set("some-key", "some-value", {
  sameSite: true
}) 
```

**Reply by oyerindedaniel** [reply]
Thanks, your solution works

**Reply by oyerindedaniel** [reply]
i am setting my cookie in after login like this in my login router 

```
ctx.resHeaders.set(
        'set-cookie',
        cookie.serialize('token', 'daniel--sss', {
          httpOnly: true,
          secure: Boolean(Number(process.env.COOKIE_SECURE ?? 0)),
          sameSite: 'lax'
          // domain: 'http://localhost:3000'
        })
      );
```


but when i try accessing it on trpc server request the cookie is empty this is my 
server.ts
// not seeing the set cookie here. I am sure it is sent because on route handler request in my next route i have access to it

```
import "server-only";

import { cache } from "react";
import { cookies, headers } from "next/headers";

import { createCaller, createTRPCContext } from "@repo/api";

/**
 * This wraps the `createTRPCContext` helper and provides the required context for the tRPC API when
 * handling a tRPC call from a React Server Component.
 */
const createContext = cache(() => {

console.log(cookies()) // not seeing the set cookie here. I am sure it is sent because on route handler request in my next route i have access to it
  const heads = new Headers(headers());
  heads.set("cookies", cookies.toString());

  return createTRPCContext({
    headers: heads,
  });
});

export const api = createCaller(createContext);
```

and this is my context
```
export const createTRPCContext = async (opts: {
  headers: Headers;
  resHeaders?: Headers;
}) => {
  const { headers } = opts;
  // const session = await db.

  return {
    db,
    ...opts,
  };
};
```

**Reply by IDJGILL** [reply]
Don't use resHeaders to set cookies:

```ctx.resHeaders.set(
        'set-cookie',
        cookie.serialize('token', 'daniel--sss', {
          httpOnly: true,
          secure: Boolean(Number(process.env.COOKIE_SECURE ?? 0)),
          sameSite: 'lax'
          // domain: 'http://localhost:3000'
        })
      );
  
Use Built-In cookies function in next/headers
import { cookies } from "next/headers"

// In your procedure
cookie.set("some-key", "some-value", {
  sameSite: true
})```

**Reply by oyerindedaniel** [reply]
thanks

**Comment by lezuber** [comment]
omfg thank you sooo much!!!
I just went away from unstable_httpBatchStreamLink! No other changes to the Contextprovider were needed!
